assets/src/edit-story/app/history/test/reducer.js

```diff
@@ -0,0 +1,273 @@
+    });
+
+    it('should not dynamically set the replay if changed happened on more than one page', () => {
+      const requestedState = {
+        current: 1,
+        pages: [
+          { id: 1, foo: 'bar' },
+          { id: 2, foo: 'bar' },
+        ],
+      };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [
+          {
+            current: 2,
+            pages: [
+              { id: 1, foo: 'foo' },
+              { id: 2, foo: 'foo' },
+            ],
+          },
+          requestedState,
+        ],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.requestedState).toBeNull();
+      expect(result.offset).toStrictEqual(1);
+    });
+
+    it('should not dynamically set the replay if a page was added', () => {
+      const requestedState = { current: 1, pages: [{ id: 1, foo: 'bar' }] };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [
+          {
+            current: 2,
+            pages: [
+              { id: 2, foo: 'foo' },
+              { id: 1, foo: 'bar' },
+            ],
+          },
+          requestedState,
+        ],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.requestedState).toBeNull();
+      expect(result.offset).toStrictEqual(1);
+    });
+  });
+
+  describe('clearHistory', () => {
+    it('should set initial state when clearing history', () => {
+      const initialState = {
+        offset: 1,
+        requestedState: { id: 1 },
+        entries: [{ id: 1 }],
+        versionNumber: 1,
+      };
+
+      const result = reducer(initialState, {
+        type: CLEAR_HISTORY,
+      });
+
+      expect(result).toMatchObject(EMPTY_STATE);
+    });
+  });
+
+  describe('replay', () => {
+    it('should set the correct requestedState when replaying', () => {
+      const replay = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 2 }, replay],
+      };
+
+      const result = reducer(initialState, {
+        type: REPLAY,
+        payload: 1,
+      });
+
+      expect(result.requestedState).toStrictEqual(replay);
+    });
+
+    it('should set the correct versionNumber when replaying', () => {
+      const replay = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        versionNumber: 4,
+        offset: 3,
+        entries: [{ id: 2 }, replay],
+      };
+
+      const result1 = reducer(initialState, {
+        type: REPLAY,
+        payload: 2,
+      });
+
+      expect(result1.versionNumber).toStrictEqual(5);
+
+      const result2 = reducer(initialState, {
+        type: REPLAY,
+        payload: 5,
+      });
+      expect(result2.versionNumber).toStrictEqual(2);
```

Comment:
Why is version number reduced here? Do we use the version number like that?

Reply:
Yes, the version number is adjusted when undo/redo happens -- otherwise, the `versionNumber` would not be really useful. The `versionNumber` was added in #877, however the #877 was not a complete implementation -- it did not consider the undo/redo at all which still allowed the user leave without notice from unsaved state. `versionNumber` was adjusted in #1361 to also consider undo/redo.

Reply:
Isn't `versionNumber` only used to check, if there's a new version of history? It's simple a _"if version number changed, something has changed"_ check?

- Is it ever used as a number anywhere?
- Who checks the actual value of it?
- Does anyone check if it's bigger than, lower than or equal to something anywhere?

Reply:
> Isn't versionNumber only used to check, if there's a new version of history? It's simple a "if version number changed, something has changed" check?

It was like that, however, it wasn't working as expected when using undo-redo. For example.
1. User does something
2. User Saves
3. User clicks "Undo"

-> It used to show as if there were no changes with the original implementation in #877 since with undo and redo the version number was not adjusted. However, actually there are new changes -- compared to the saved version.

> Does anyone check if it's bigger than, lower than or equal to something anywhere?

Yes:
https://github.com/google/web-stories-wp/blob/37fa4af305141754fc34271a1852ce6b9487f538/assets/src/edit-story/app/history/historyProvider.js#L53-L55